agent/immortalstreams/backedpipe/backed_pipe.go

```diff
@@ -0,0 +1,290 @@
+package backedpipe
+
+import (
+	"context"
+	"io"
+	"sync"
+
+	"golang.org/x/sync/singleflight"
+	"golang.org/x/xerrors"
+)
+
+var (
+	ErrPipeClosed             = xerrors.New("pipe is closed")
+	ErrPipeAlreadyConnected   = xerrors.New("pipe is already connected")
+	ErrReconnectionInProgress = xerrors.New("reconnection already in progress")
+	ErrReconnectFailed        = xerrors.New("reconnect failed")
+	ErrInvalidSequenceNumber  = xerrors.New("remote sequence number exceeds local sequence")
+	ErrReconnectWriterFailed  = xerrors.New("reconnect writer failed")
+)
+
+const (
+	// Default buffer capacity used by the writer - 64MB
+	DefaultBufferSize = 64 * 1024 * 1024
+)
+
+// ReconnectFunc is called when the BackedPipe needs to establish a new connection.
+// It should:
+// 1. Establish a new connection to the remote side
+// 2. Exchange sequence numbers with the remote side
+// 3. Return the new connection and the remote's current sequence number
+//
+// The writerSeqNum parameter is the local writer's current sequence number,
+// which should be sent to the remote side so it knows where to resume reading from.
+//
+// The returned readerSeqNum should be the remote side's current sequence number,
+// which indicates where the local reader should resume from.
+type ReconnectFunc func(ctx context.Context, writerSeqNum uint64) (conn io.ReadWriteCloser, readerSeqNum uint64, err error)
+
+// BackedPipe provides a reliable bidirectional byte stream over unreliable network connections.
+// It orchestrates a BackedReader and BackedWriter to provide transparent reconnection
+// and data replay capabilities.
+type BackedPipe struct {
+	ctx         context.Context
+	cancel      context.CancelFunc
+	mu          sync.RWMutex
+	reader      *BackedReader
+	writer      *BackedWriter
+	reconnectFn ReconnectFunc
```

Comment:
In order to handle both initial connections and reconnections, the thing that handles the connections is going to need to be stateful. On the client side it needs to post a new immortal stream the first time, and then store the name/ID of the stream for re-dialing.  Instead of making this a function, make it an interface type with the reconnection function on it.

It's much easier to understand this kind of thing because you can ask your IDE to show you all implementations of an interface type.

It also avoids the awkward closure that you define up the stack to handle the statefulness. It can be a regular function on the struct type rather than an anonymous closure over it. Also makes stack traces way more readable.